docs/topics/coroutines.rst

```diff
@@ -0,0 +1,103 @@
+
+.. warning:: :mod:`asyncio` support in Scrapy is experimental. Future Scrapy
+             versions may introduce related API and behavior changes without a
+             deprecation period or warning.
+
+.. _coroutine-support:
+
+Supported callables
+===================
+
+The following callables may be defined as coroutines using ``async def``, and
+hence use coroutine syntax (e.g. ``await``, ``async for``, ``async with``):
+
+-   :class:`~scrapy.http.Request` callbacks.
+
+    The following are known caveats of the current implementation that we aim
+    to address in future versions of Scrapy:
+
+    -   The callback output is not processed until the whole callback finishes.
+
+        As a side effect, if the callback raises an exception, none of its
+        output is processed.
+
+    -   Because `asynchronous generators were introduced in Python 3.6`_, you
+        can only use ``yield`` if you are using Python 3.6 or later.
+
+        If you need to output multiple items or requests and you are using
+        Python 3.5, return an iterable (e.g. a list) instead.
+
+-   The :meth:`process_item` method of
+    :ref:`item pipelines <topics-item-pipeline>`.
+
+-   The
+    :meth:`~scrapy.downloadermiddlewares.DownloaderMiddleware.process_request`,
+    :meth:`~scrapy.downloadermiddlewares.DownloaderMiddleware.process_response`,
+    and
+    :meth:`~scrapy.downloadermiddlewares.DownloaderMiddleware.process_exception`
+    methods of
+    :ref:`downloader middlewares <topics-downloader-middleware-custom>`.
+
+-   :ref:`Signal handlers that support deferreds <signal-deferred>`.
+
+.. _asynchronous generators were introduced in Python 3.6: https://www.python.org/dev/peps/pep-0525/
+
+Usage
+=====
+
+There are several use cases for coroutines in Scrapy. Code that would
+return Deferreds when written for previous Scrapy versions, such as downloader
+middlewares and signal handlers, can be rewritten to be shorter and cleaner::
+
+    class DbPipeline():
+        def _update_item(self, data, item):
+            item['field'] = data
+            return item
+
+        def process_item(self, item, spider):
+            dfd = db.get_some_data(item['id'])
+            dfd.addCallback(self._update_item, item)
+            return dfd
+
+becomes::
+
+    class DbPipeline():
+        async def process_item(self, item, spider):
+            item['field'] = await db.get_some_data(item['id'])
+            return item
+
+Coroutines may be used to call asynchronous libraries that use either Deferreds
+or coroutines::
+
+    async def parse_with_deferred(self, response):
+        additional_response = await treq.get('https://additional.url')
+        additional_data = await treq.content(additional_response)
+
+    async def parse_with_asyncio(self, response):
+        async with aiohttp.ClientSession() as session:
+            async with session.get('https://additional.url') as additional_response:
+                additional_data = await r.text()
```

Comment:
Could you please make it a more complete example, with Spider definition? It could also make sense to show that you can output an item based on the data we're getting.

Reply:
I can add the class definition line, but a complete spider example would be too long I think. Same about the callback methods: if we want complete examples we would need to do something with both the original response and the additional one, that would make them too long, wouldn't it?

Reply:
It can be a comment which says that, not a real code; added a suggestion.